python/pyarrow/tests/test_array.py

```diff
@@ -4012,3 +4012,91 @@ def test_swapped_byte_order_fails(numpy_native_dtype):
     # Struct type array
     with pytest.raises(pa.ArrowNotImplementedError):
         pa.StructArray.from_arrays([np_arr], names=['a'])
+
+
+def test_non_cpu_array():
+    cuda = pytest.importorskip("pyarrow.cuda")
+    ctx = cuda.Context(0)
+
+    data = np.arange(4, dtype=np.int32)
+    validity = np.array([True, False, True, False], dtype=np.bool_)
+    cuda_data_buf = ctx.buffer_from_data(data)
+    cuda_validity_buf = ctx.buffer_from_data(validity)
+    arr = pa.Array.from_buffers(pa.int32(), 4, [None, cuda_data_buf])
+    arr2 = pa.Array.from_buffers(pa.int32(), 4, [None, cuda_data_buf])
+    arr_with_nulls = pa.Array.from_buffers(
+        pa.int32(), 4, [cuda_validity_buf, cuda_data_buf])
+
+    # Supported
+    arr.validate()
+    arr.validate(full=True)
+    assert arr.offset == 0
+    assert arr.buffers() == [None, cuda_data_buf]
+    assert arr.device_type == pa.DeviceAllocationType.CUDA
+    assert arr.is_cpu is False
+    assert len(arr) == 4
+    assert arr.slice(2, 2) == pa.RecordBatch.from_arrays([
+        ctx.buffer_from_data(data[2:])
+    ])
```

Comment:
This is failing, because the RecordBatch.from_arrays is failing (needs names), but regardless of that, we also don't need a RecordBatch but an Array ;) 
But the problem is that even when you reconstruct an expected Array from a sliced buffer, the `==` doesn't work for non-CPU data ...

For now, you could maybe just test that the offset if set on the result? 

```suggestion
    assert arr.slice(2, 2).offset == 2
```

Longer term we could test this by copying the result to CPU and verify equality there (but this needs https://github.com/apache/arrow/pull/42223)

Reply:
Thank you! Great suggestion :)